src/rendering/luaconsole.cpp

```diff
@@ -886,12 +688,589 @@ void LuaConsole::setCommandInputButton(Key key) {
+
+    auto containsNonAscii = [](const std::filesystem::path& p) {
+        const std::u8string s = p.generic_u8string();
+        for (auto it = s.rbegin(); it != s.rend(); it++) {
+            if (static_cast<unsigned char>(*it) > 0x7F) {
+                return true;
+            }
+        }
+        return false;
+    };
+
+    std::vector<std::string> entries;
+    for (const std::filesystem::path& entry : suggestions) {
+        // Filter paths that contain non-ASCII characters
+        if (containsNonAscii(entry)) {
+            continue;
+        }
+
+        entries.push_back(entry.string());
+    }
+
+    _autoCompleteState.suggestions = std::move(entries);
+    _autoCompleteState.input = userTypedPath;
+
+    if (pathEnd != std::string::npos) {
+        // There is something after the path so we want to insert inbetween
+        _autoCompleteState.insertPosition = contextStart + pathEnd;
+    }
+    else {
+        // There is nothing after the path so we render normally at the end
+        _autoCompleteState.insertPosition = currentCommand.size();
+    }
+
+    return true;
+}
+
+void LuaConsole::gatherFunctionSuggestions(size_t contextStart) {
+    const std::string currentCommand = _commands[_activeCommand];
+    // Get a list of all the available commands
+    std::vector<std::string> allCommands = global::scriptEngine->allLuaFunctions();
+    std::sort(allCommands.begin(), allCommands.end());
+
+    _autoCompleteState.suggestions = std::move(allCommands);
+
+    std::string possibleFunction = currentCommand.substr(contextStart);
+    // Find the nearest parenthesis if any exists, which marks the end of the
+    // function. Otherwise the rest of the string is assumed to be part of
+    // the function
+    size_t functionEnd = possibleFunction.find_first_of("()");
+    _autoCompleteState.input = possibleFunction.substr(0, functionEnd);
+
+    if (functionEnd != std::string::npos) {
+        // There is something after the function so we want to insert inbetween
+        _autoCompleteState.insertPosition = contextStart + functionEnd;
+    }
+    else {
+        // There is nothing after the path so we render normally at the end
+        _autoCompleteState.insertPosition = currentCommand.size();
+    }
+}
+
+void LuaConsole::filterSuggestions() {
+    auto normalize = [this](const std::string& s) {
+        std::string out = s;
+
+        if (_autoCompleteState.context == Context::Function) {
+            out = ghoul::toLowerCase(out);
+        }
+
+#ifdef WIN32
+        // On Windows, file paths are generally case-insensitive. For example,
+        // "C:/User/Desktop/Foo" refers to the same location as "c:/user/desktop/foo"
+        // Normalize paths to lowercase so they are treated equivalently
+        if (_autoCompleteState.context == Context::Path) {
+            out = ghoul::toLowerCase(sanitizeInput(out));
+        }
+#endif // WIN32
+
+        return out;
+    };
+
+    // We're only interested in unique matches
```

Comment:
```suggestion
    // We're only interested in unique matches, and want them sorted alphabetically, 
    // so use a set for the results
```

Reply:
However, there is no need to sort or remove duplicates until the very end, and since you create a vector from this set in the end I would recommed just using a std::vector right away. See other suggestion

That would make the code more explicit and hence easier to read/understand, and probably also be a bit faster :)